drivers/leds/ws2812.c

```diff
@@ -0,0 +1,475 @@
+            }
+
+          offset++;
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: ws2812_open
+ *
+ * Description:
+ *   This function is called whenever the WS2812 device is opened.
+ *
+ ****************************************************************************/
+
+static int ws2812_open(FAR struct file *filep)
+{
+  return OK;
+}
+
+/****************************************************************************
+ * Name: ws2812_close
+ *
+ * Description:
+ *   This routine is called when the WS2812 device is closed.
+ *
+ ****************************************************************************/
+
+static int ws2812_close(FAR struct file *filep)
+{
+  return OK;
+}
+
+/****************************************************************************
+ * Name: ws2812_read
+ ****************************************************************************/
+
+static ssize_t ws2812_read(FAR struct file *filep, FAR char *buffer,
+                           size_t buflen)
+{
+  return -ENOSYS;
+}
+
+/****************************************************************************
+ * Name: ws2812_write
+ *
+ * Description:
+ *   This routine is called when writing to the WS2812 device. Data buffer
+ *   should be an array of 32bit values holding 24bits of color information
+ *   in host byte ordering 0x**rrggbb.
+ *
+ ****************************************************************************/
+
+static ssize_t ws2812_write(FAR struct file *filep, FAR const char *buffer,
+                            size_t buflen)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct ws2812_dev_s *priv = inode->i_private;
+  FAR uint8_t *tx_pixel;
+  FAR uint32_t *pixel_buf = (FAR uint32_t *)buffer;
+  size_t cur_led;
+  size_t start_led;
+  size_t end_led;
+  size_t written = 0;
+
+  if (buffer == NULL)
+    {
+      lederr("ERROR: Buffer is null\n");
+      set_errno(EINVAL);
+      return -1;
+    }
+
+  /* We need at least one display, so 1 byte */
+
+  if (buflen < 1)
+    {
+      lederr("ERROR: You need to control at least 1 LED!\n");
+      set_errno(EINVAL);
+      return -1;
+    }
+
+  if ((buflen % WS2812_RW_PIXEL_SIZE) != 0)
+    {
+      lederr("ERROR: LED values must be 24bit packed in 32bit\n");
+      set_errno(EINVAL);
+      return -1;
+    }
+
+  nxsem_wait(&priv->exclsem);
+
+  start_led = filep->f_pos / WS2812_RW_PIXEL_SIZE;
+  tx_pixel = priv->tx_buf + WS2812_RST_CYCLES + \
+             start_led * WS2812_BYTES_PER_LED;
+
+  end_led = start_led + (buflen / WS2812_RW_PIXEL_SIZE) - 1;
+  ledinfo("Start: %d End: %d\n", start_led, end_led);
+
+  if (end_led  > (priv->nleds -1))
+    {
+      end_led = priv->nleds - 1;
+    }
+
+  for (cur_led = start_led; cur_led <= end_led; cur_led++)
+    {
+      ws2812_pack(tx_pixel, *pixel_buf & 0xffffff);
+      pixel_buf++;
+      tx_pixel += WS2812_BYTES_PER_LED;
+      written += WS2812_RW_PIXEL_SIZE;
+    }
+
```

Comment:
Do you need to lock the SPI bus and then call ws2812_configspi here?

Reply:
See the lock here https://github.com/apache/incubator-nuttx/pull/1797/files#diff-c94b56b8563460b40fec75e2c397c68cR453

You cannot really share the bus if this device is connected to it so I just lock it on init so it should be very obvious something is wrong if you try to share.

Reply:
Ok, Then why not hold the lock `SPI_LOCK(priv->spi, true);` then do the init `ws2812_configspi(priv->spi);` ? 
Also maybe add a disclaimer to not use this on a common bus to the [Kconfig]( https://github.com/apache/incubator-nuttx/pull/1797/files#diff-a616198bfed58f9e3b4f936ee5284b1cR97)

Reply:
Done.

Reply:
What about ensuring the driver is holding the SPI lock before changing the SPI settings?

Reply:
@davids5 sorry about that I did not `git add` the file... I was very confused, that is the change I made.
https://github.com/apache/incubator-nuttx/pull/1797/files#diff-c94b56b8563460b40fec75e2c397c68cR136

Reply:
No worries. All good now! Shall we merge it?

Reply:
Sounds good to me.  I have thoughts to extend it with some IOCTL, but that would be in a future PR.